### DISCUSSION TITLE
Cross-relational validations

### DISCUSSION THREAD

**Discussion by viters** [directus/directus#16852] [Feature Requests]
## Idea

It would be useful if editing `YY` collection object through `XX` collection object form (`XX` contains multiple `YY`) was traceable on hooks function level. It is really inconvenient (or impossible) right now to write cross-relational validation `filter` hooks and I think additional context is required.

```js
filter('XX.items.update', (payload, meta, ctx) => {
  // ...
});

filter('YY.items.update', (payload, meta, ctx) => {
  // if YY update was result of XX update, there is no trace of that here
});
```

## Example use case

You have `events` collection and `workshops` collection. `event` has multiple `workshops`.
Both `event` and `workshop` have `startDate` and `endDate`.

I was trying to verify a simple validation requirement: `workshops` should be within `event`'s dates. For example, if `event` is taking place from 12 Dec to 14 Dec, setting `workshop` to 15 Dec to 16 Dec should fail. Also, if `event` already contains `workshop` item from 15 Dec to 16 Dec, modifying `event`'s time range to not contain 15-16 Dec should fail as well.

The validation should work in all cases: 
- modifying `workshop` directly
- modifying `event` through `workshop` form (relational)
- modifying `event` directly
- modifying `workshop` through `event` form (relational)

Consider a scenario, where I update `event`'s dates and related `workshop` dates within the same form. When I click save, the flow is:
1. filter `events.items.update` called, that receives both info about changes to `event` and `workshop`
2. filter `workshops.items.update` called, that receives **only** info about changes to `workshop`
3. **only** if all `filters` pass, `event` and `workshop` is saved

## Current solution

There is no reliable current solution to that use case.

## Possible solution 1

Allow user to set custom values that are propagated through context of hooks run. Similar to AsyncLocalStorage.

```js
filter('events.items.update', (payload, meta, ctx) => {
  if (payload.workshops != null && (payload.startDate != null || payload.endDate != null)) {
    // YY.items.update will be called
   ctx.params = { type: 'eventDatesUpdate', event: { startDate: payload.startDate, endDate: payload.endDate } };
  }
});

filter('workshops.items.update', (payload, meta, ctx) => {
  if (ctx.params?.type === 'eventDatesUpdate') {
    // run validation against ctx.params
  }
});
```

## Possible solution 2

Change the order of operations in `ItemsService`, in a way that `knex` transaction contains applied changes from previously validated steps.

```js
filter('events.items.update', (payload, meta, ctx) => {
  const workshops = ctx.database('workshops').where...
  // ctx is root knex or transaction depending on situation
  // that should result in updated workshops if the `workshop` object is the source of update
});

filter('YY.items.update', (payload, meta, ctx) => {
  const event = ctx.database('events').where...
  // ctx is root knex or transaction depending on situation
  // that should result in updated event if the `event` object is the source of update
});
```

Right now such solution will not work, as every filter is run before any changes are applied to transaction.

## Possible solution 3

Validation hook API. TBD.

**Comment by rijkvanzanten** [comment]
Wondering if we could put the revision from the "parent" object as part of the `meta` information of the current hook execution 🤔 That would potentially help here as well

**Reply by viters** [reply]
I have been tinkering with @br41nslug and theres one more issue that I guess it would be useful to address at the same time. Your solution should allow to perform that validation, but it will still be cumbersome.

I've written an example of validation for Events > Workshops > Participants relation. Events have multiple workshops, and each workshop has participants. There should be no more participants than `guestLimit` on event.

```js
module.exports = ({filter}) => {
  filter('events.items.update', async (payload, meta, { database }) => {
    console.log('events.items.update', JSON.stringify(payload));

    if (payload.guest_limit != null) {
      const participantsSum = (await database('workshops')
        .join('participants', 'workshops.id', '=', 'participants.workshop_id')
        .select('workshops.id as id')
        .count('participants.id as participants')
        .where('workshops.event_id', meta.keys[0])
        .groupBy('workshops.id')).reduce((prev, next) => prev + (+next.participants), 0);

      if (participantsSum > payload.guest_limit) {
        throw new Error(`EVENT VALIDATION: participantsSum ${participantsSum} > payload.guest_limit ${payload.guest_limit}`)
      }
    }
  });

  /*
    It is tedious to write the validation from different perspectives
    considering that it verifies the exact same thing.
   */

  filter('workshops.items.update', async (payload, meta, { database }) => {
    console.log('workshops.items.update', JSON.stringify(payload));

    if (payload.participants != null) {
      const event = await database('workshops').where('workshops.id', meta.keys[0]).join('events', 'events.id', '=', 'workshops.event_id').first();

      const resultParticipantsSum = (await database('workshops')
        .join('participants', 'workshops.id', '=', 'participants.workshop_id')
        .select('workshops.id as id')
        .count('participants.id as participants')
        .where('workshops.event_id', event.id)
        .groupBy('workshops.id')).reduce((prev, next) => prev + (+next.participants), 0)
        // Add participants created via "Create new"
        + payload.participants.create.length
        // Add participants attached to this workshop
        + payload.participants.update.filter(a => a.workshop_id === meta.keys[0]).length
        // Subtract participants detached from this workshop
        - payload.participants.update.filter(a => a.workshop_id !== meta.keys[0]).length
        // Subtract deleted participants (probably not possible via UI)
        - payload.participants.delete.length;

      if (resultParticipantsSum > event.guest_limit) {
        throw new Error(`WORKSHOPS VALIDATION: resultParticipantsSum ${resultParticipantsSum} > event.guest_limit ${event.guest_limit}`)
      }
    }
  });

  filter('participants.items.update', (payload, meta) => {
    console.log('participants.items.update', JSON.stringify(payload));
    // This probably needs the validation as well, from participant perspective
  });
};
```

This does not work at the moment at all. The problem is that even if it would, it is inconvenient to write and maintain. Right now you have to write the validation **from each relations perspective**. I think there should be a way to write the validation only once.

Example idea:

```js
module.exports = ({validate}) => {
  validate('events,workshops,participants', async (transaction) => {
    // here all changes are already applied to transaction,
    // it does not matter where the change originated from
    // I can run SQL queries to verify if everything is all right and throw if not
  };
};

**Comment by viters** [comment]
@br41nslug 
Have you maybe had time to think about that problem or maybe you have discussed it internally?

**Reply by br41nslug** [reply]
The question of nested validation here is a tricky one. Sorry but can't say there has been a lot of movement on this so far.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
